### PR TITLE
Migrate line and progress state to unencrypted cookies.

### PR DESCRIPTION
**Pull request by philbogle**
Move session[:lines] and session[:progress] to JSON-encoded state in cookies[:lines] and cookies[:progress] that can be read from Javascript.

This PR is a step on the way to being able to cache HOC pages in the CDN across users.

I wrote unit tests to verify that state is correctly migrated and also manually verified that my state was preserved when I updated and that cookies[:progress] and cookies[:lines] has the correct values.

While deploying this, we should temporarily on soft-stickiness on the load balancer so that users whose state has been moved to unencrypted cookies will continue to talk to a frontend that knows how to deal with those cookies.  Soft-stickiness can be turned off as soon as all frontends have updated.

We shouldn't push this to production until my preceeding user_session_helpers PR has gone live in production. That CL includes reverse migration logic so that even if we do a rollback of the release including this CL, user progress will be preserved.

### CODE REVIEW

**Comment by philbogle** [comment]
@mehalshah And here is the follow-on pull request.

**Comment by philbogle** [comment]
@mehals You probably want to look at the diff against user_session_helpers to avoid seeing changes again that you have seen already:

https://github.com/code-dot-org/code-dot-org/compare/user_session_helpers...user_progress_cookies

**Comment by philbogle** [comment]
Corrected by comment to refer to mehalshah rather than mehals.

File: dashboard/lib/client_state.rb [hunk 1]
```diff
@@ -13,9 +13,8 @@ def initialize(session, cookies)
 
   # Resets all client state (level progress, lines of code, videos seen, etc.)
   def reset
-    # Also reset unencrypted cookies in case we are in a rolled-back state.
-    session[:lines] = cookies[:lines] = nil
-    session[:progress] = cookies[:progress] = nil
+    cookies[:lines] = nil
```
Comment (mehalshah) [inline dashboard/lib/client_state.rb:16]: If your intention is to set every value of session and cookies to nil, you can shorten this with

[cookies, session].each { |a| a.each { |k, _| a[k] = nil}}

Comment (philbogle) [inline dashboard/lib/client_state.rb:16]: This class is only intended manage state that is potentially needed on the client;  I didn't want to whack other session or cookie state that people might add for other reasons.  (Suppose, for example, someone added session state for tracking the number of times the user has visited the site-- you wouldn't want that to get whacked when client progress was reset.)

Comment (mehalshah) [inline dashboard/lib/client_state.rb:16]: Okay, ignore this comment then

File: dashboard/lib/client_state.rb [hunk 2]
```diff
@@ -118,30 +117,22 @@ def add_callout_seen(callout_key)
 
   private
 
-  # Migrates cookies[:progress] back to session[:progress] in case we end up rolling
-  # back the release that does the forward migration. Except in that case, this code
-  # does nothing becaue we don't otherwise use cookies[:progress] and cookies[:lines].
-  def undo_cookie_migration_on_rollback
-    if cookies[:progress].present? && session[:progress].blank?
+  def progress_hash
```
Comment (laurelfan) [inline dashboard/lib/client_state.rb:120]: did you know that in ruby you can rescue an entire method without a begin/end block?

```
  def progress_hash
    migrate_cookies
    if cookies[:progress]
      JSON.parse(cookies[:progress])
    else
      {}
    end
  rescue JSON::ParserError
      {}
  end
```

(your choice whether you like the ruby-ism of not using return when not necessary for control flow...)

File: dashboard/lib/client_state.rb [hunk 3]
```diff
@@ -118,30 +117,22 @@ def add_callout_seen(callout_key)
...
     end
-
-    if cookies[:lines].present? && session[:lines].blank?
-      session[:lines] = cookies[:lines].to_i
-      cookies[:lines] = nil
-    end
+    return {}
   end
 
   # Migrates session state to unencrypted cookies.  This is currently used in tests only
```
Comment (laurelfan) [inline dashboard/lib/client_state.rb]: "tests only" comment does not apply anymore?

File: dashboard/lib/client_state.rb (hunk 2 above)
Comment (philbogle) [inline dashboard/lib/client_state.rb:120]: Done.

File: dashboard/lib/client_state.rb (hunk 3 above)
Comment (philbogle) [inline dashboard/lib/client_state.rb]: Done.

**Comment by philbogle** [comment]
Replied to Laurel's comments.

**Comment by laurelfan** [comment]
lgtm

**Pull request merged**